### PR TITLE
Implements highlighting of unicode characters.

### LINKED ISSUE #136437: Highlight confusable Unicode characters
*updated*

UNTRUSTED WORKSPACES:
* there's a feature on by default that non basic ASCII is highlighted --> boolean
    * defaults to true (false in trusted)
    * "editor.unicodeHighlight.nonBasicASCII"

TRUSTED WORKSPACES:
* highlight unexpected invisibles code points (on by default) ---> boolean
    * defaults to true
    * some smartness if they appear in emojis
    * "editor.unicodeHighlight.invisibleCharacters"
    * does not highlight characters that appear in the vscode-loc translation files of the current language.

* highlight unexpected confusable code points ---> boolean (linting feature)
    * on the worker side, have a set of dangerous chars (includes Cyrilic!) and return all ranges with them
    * create a set of 1000 characters ('a' cyrilic) 
       * FIND false positives in repos nodejs/linux -- 900 are problematic or 100 or 10?
       * EXCEPTIONS (remove things from this set ("false positives"))
       * never highlight comments or strings (filter on the UI side, needs some work with background tokenization?)
       * comments, strings, 
       * if your display language contains those
       * or if your keyboard layout produced them
       * Intl (system defined preferences?), like env.LOCALE
       * intersect with vscode-loc
    * defaults to true except for markdown, plain text, latex
    * "editor.unicodeHighlight.confusables"

-> try to start with NO setting to define further "exceptions" for confusables

### PR DESCRIPTION
**Pull request by hediet**
This PR implements #136437

Some unicode characters are highlighted:

image

There is an hover to explain why:

image

"Adjust Settings" will open a quick pick:
image

## New Settings
* `editor.unicodeHighlight.excludedCharacters: string[]`
* `editor.unicodeHighlight.invisibleCharacters: boolean`
* `editor.unicodeHighlight.nonBasicASCII: boolean`
* `editor.unicodeHighlight.ambiguousCharacters: boolean`
* `editor.unicodeHighlight.includeComments: boolean`

For now, everything defaults to `false` or `[]`, but it is planned to use the trust state to configure the defaults:
* Untrusted:
    * `editor.unicodeHighlight.excludedCharacters: []`
    * `editor.unicodeHighlight.invisibleCharacters: true`
    * `editor.unicodeHighlight.nonBasicASCII: true`
    * `editor.unicodeHighlight.ambiguousCharacters: true`
    * `editor.unicodeHighlight.includeComments: true`
* Trusted:
    * `editor.unicodeHighlight.excludedCharacters: []`
    * `editor.unicodeHighlight.invisibleCharacters: true`
    * `editor.unicodeHighlight.nonBasicASCII: false`
    * `editor.unicodeHighlight.ambiguousCharacters: true`
    * `editor.unicodeHighlight.includeComments: false`

## New Theme Colors
* `editorUnicodeHighlight.border` to control the border of the highlight box (default red)

## Invisible characters

An unicode character is invisible if rendering it with a black font on a white canvas leaves the canvas white.

## Ambiguous characters

A character is ambiguous if it is confusable with a basic ascii character and if it does not appear in vscode-loc for the currently selected locale.

## Follow ups
* Don't highlight invisible characters in emojis
* Adjust the defaults
* Evaluate if invisible characters that appear in vscode-loc should be excluded (warning: this includes zero width space etc.). Note, that for trusted workspace, it is not planned to highlight unicode characters in comments
* [ ] Never highlight characters in markdown/text files

### CODE REVIEW

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+		}
+		return result;
+	}
+
+	public renderHoverParts(hoverParts: MarkdownHover[], fragment: DocumentFragment, statusBar: IEditorHoverStatusBar): IDisposable {
+		return renderMarkdownHovers(hoverParts, fragment, this._editor, this._hover, this._modeService, this._openerService);
+	}
+}
+
+function escapedMarkdownInlineCode(code: string): string {
```
Comment (alexdima) [inline src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:369]: This function appears overengineered to me. It looks like it is only called with a single code point as the value of `code`, yet its implementation deals with the complex case of escaping arbitrary long sequences of backticks. A more simple function could be changed to simply cater for `` code === '`' `` and hard code the return value of ``` `` ` `` ``` in that case.

```
function renderCodePointAsInlineCode(codePoint: number): string {
	if (codePoint === CharCode.BackTick) {
		return '`` ` ``';
	}
	return '`' + String.fromCodePoint(codePoint) + '`';
}
```

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+					reason = nls.localize(
+						'unicodeHighlight.characterIsAmbiguous',
+						'The character {0} could be confused with the more common character {1}.',
+						codePointStr,
+						formatCodePoint(highlightInfo.reason.confusableWith.codePointAt(0)!)
+					);
+					break;
+
+				case 'invisible':
+					reason = nls.localize(
```
Comment (alexdima) [inline src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:332]: Bi-Di Control characters should never be really rendered in the hover:

image

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+	return UnicodeTextModelHighlighter.computeUnicodeHighlightReason(char, options);
+}
+
+const DECORATION_HIDE_IN_COMMENTS = ModelDecorationOptions.register({
+	description: 'unicode-highlight',
+	stickiness: TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges,
+	className: 'unicode-highlight',
+	showIfCollapsed: true,
+	overviewRuler: {
+		color: themeColorFromId(overviewRulerFindMatchForeground),
```
Comment (alexdima) [inline src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:391]: We could define special colors for this, I just used the find match colors in a hurry.

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+		position: OverviewRulerLane.Center
+	},
+	minimap: {
+		color: themeColorFromId(minimapFindMatch),
```
Comment (alexdima) [inline src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:395]: same here with using the find match color.

**Review by alexdima — CHANGES_REQUESTED** [review]
:+1: Looks pretty good! I have finished looking at the code, I will spend some time looking at files and will try to write more feedback after that.

File: src/vs/editor/common/config/editorOptions.ts [hunk 5]
```diff
@@ -3243,6 +3259,121 @@ class EditorScrollbar extends BaseEditorOption<EditorOption.scrollbar, InternalE
...
+			}
+		);
+	}
+
+	public validate(_input: any): InternalUnicodeHighlightOptions {
+		if (!_input || typeof _input !== 'object') {
+			return this.defaultValue;
+		}
+		const input = _input as IUnicodeHighlightOptions;
+		return {
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/config/editorOptions.ts:3342]: At this location, the trust state is unknown. Thus, we cannot derive the default from it.
As a workaround, the value is `undefined` to indicate that it should take the default once the trust state can be obtained.

File: src/vs/editor/common/config/editorOptions.ts [hunk 6]
```diff
@@ -3243,6 +3259,121 @@ class EditorScrollbar extends BaseEditorOption<EditorOption.scrollbar, InternalE
...
+		};
+
+		super(
+			EditorOption.unicodeHighlighting, 'unicodeHighlight', defaults,
+			{
+				[unicodeHighlightConfigKeys.nonBasicASCII]: {
+					restricted: true,
+					type: 'boolean',
+					default: defaults.nonBasicASCII,
+					description: nls.localize('unicodeHighlight.nonBasicASCII', "Controls whether all non-basic ASCII characters are highlighted. Only characters between U+20 and U+7E, tab, line-feed and carriage-return are considered basic ASCII.")
```
[the remark this answers is not in the document]
Comment (alexdima) [reply src/vs/editor/common/config/editorOptions.ts:3307]: Also seen in the hover. I think `U+00AD` might look better:

image

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+					break;
+			}
+
+			const adjustSettingsArgs: ShowExcludeOptionsArgs = {
+				codePoint: codePoint,
+				reason: highlightInfo.reason.kind,
+			};
+
+			const adjustSettings = nls.localize('unicodeHighlight.adjustSettings', 'Adjust settings');
+			const contents: Array<IMarkdownString> = [{
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:354]: The hover stays when you click on the link.
I experimented with listing these actions in the hover, but they took too much space and it did not look great.
Also, I think the actions "Disable {ambiguous, invisible, non-basic ascii} characters for {markdown, text, ...} files" or "Allow all characters in this file" might make sense.

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+			}
+
+			const codePointStr = formatCodePoint(codePoint);
+
+			let reason: string;
+			switch (highlightInfo.reason.kind) {
+				case 'ambiguous':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsAmbiguous',
+						'The character {0} could be confused with the more common character {1}.',
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:325]: I don't think it is neccessary to mention the unicode spec. It is good to explicitly say "more common in source code" though.

File: src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts (hunk 2 above)
Comment (hediet) [reply src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts:333]: I think it does not make sense to render invisible characters anyway. I will just remove the `"..."`.

File: src/vs/editor/common/modes/unicodeTextModelHighlighter.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+	public static computeUnicodeHighlights(model: IUnicodeCharacterSearcherTarget, options: UnicodeHighlighterOptions, range?: IRange): Range[] {
+		const startLine = range ? range.startLineNumber : 1;
+		const endLine = range ? range.endLineNumber : model.getLineCount();
+
+		const codePointHighlighter = new CodePointHighlighter(options);
+
+		// Only check for non-basic ASCII characters
+		const regex = new RegExp(UnicodeTextModelHighlighter.NON_BASIC_ASCII_REGEX, 'g');
+		const searcher = new Searcher(null, regex);
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/modes/unicodeTextModelHighlighter.ts:25]: I added a limit for 1k "bad" characters. I don't think calling 1k times into the regexp causes a measurable performance hit. If so, I will improve that.

Also, the regex now stops only for invisible and ambiguous characters.

File: src/vs/editor/common/config/editorOptions.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/config/editorOptions.ts:3345]: I introduced a three state thing as discussed.

**Review by alexdima — APPROVED** [review]
Nice!

**Comment by alexdima** [comment]
We can add the scripts used to generate `AmbiguousCharacters` and `InvisiblesCharacters` to https://github.com/alexdima/unicode-utils such that we have a chance again in the future to update with newer Unicode versions.

**Comment by Keno** [comment]
Is there a way to turn this off by default depending on the source language. In julia source files, unicode variable names are very common and they now all have ugly red boxes around them by default:

Screenshot from 2021-11-25 22-50-13

I understand that it's possible to turn this off manually, but this makes the first-use experience much worse for users opening julia source files.

**Comment by davidanthoff** [comment]
If there is any way that we can avoid this shipping to the Julia user base on the stable release channel, I would greatly appreciate it. This is such a visible regression that would affect ten thousands of Julia users, that I’m very worried about the support request wave this would generate. We are a small team of volunteers that maintains the Julia extension, we are simply not equipped well to handle situations where something so fundamental regresses for a large fraction of our (large) user base. Is there some way that either we can disable this for Julia files before it ships (like Keno suggested), or could this be delayed from shipping until there is some other workable solution for languages like Julia that routinely use Unicode in variable names?

**Comment by hediet** [comment]
@Keno @davidanthoff let's discuss in #137924

**Comment by Ebuall** [comment]
Is it possible to exclude strings?

**Comment by chrisinmtown** [comment]
What is the expected release for this new feature?

**Comment by hediet** [comment]
Yesterday.

**Pull request merged**